src/main/scala/cognite/spark/v1/DefaultSource.scala

```diff
@@ -188,104 +188,108 @@ class DefaultSource
-      }
-      val batchSizeDefault = relation match {
-        case _: SequenceRowsRelation => Constants.DefaultSequenceRowsBatchSize
-        case _ => Constants.DefaultBatchSize
-      }
-      val batchSize = config.batchSize.getOrElse(batchSizeDefault)
-      val originalNumberOfPartitions = data.rdd.getNumPartitions
-      val idealNumberOfPartitions = config.sparkPartitions
-
-      // If we have very many partitions, it's quite likely that they are significantly uneven.
-      // And we will have to limit parallelism on each partition to low number, so the operation could
-      // take unnecessarily long time. Rather than risking this, we'll just repartition data in such case.
-      // If the number of partitions is reasonable, we avoid the data shuffling
-      val dataRepartitioned =
-        if (originalNumberOfPartitions > 50 && originalNumberOfPartitions > idealNumberOfPartitions) {
-          data.repartition(idealNumberOfPartitions)
-        } else {
-          data
-        }
-
-      dataRepartitioned.foreachPartition((rows: Iterator[Row]) => {
-        import CdpConnector.ioRuntime
 
-        val maxParallelism = config.parallelismPerPartition
-        val batches = Stream.fromIterator[IO](rows, chunkSize = batchSize).chunks
+    val relation: CdfRelation = resourceType match {
+      case AssetHierarchyBuilder.name =>
+        new AssetHierarchyBuilder(config)(sqlContext)
+      case NumericDataPointsRelation.name =>
+        new NumericDataPointsRelationV1(config)(sqlContext)
+      case StringDataPointsRelation.name =>
+        new StringDataPointsRelationV1(config)(sqlContext)
+      case EventsRelation.name =>
+        new EventsRelation(config)(sqlContext)
+      case TimeSeriesRelation.name =>
+        new TimeSeriesRelation(config)(sqlContext)
+      case AssetsRelation.name =>
+        new AssetsRelation(config)(sqlContext)
+      case FilesRelation.name =>
+        new FilesRelation(config)(sqlContext)
+      case SequenceRelation.name =>
+        new SequencesRelation(config)(sqlContext)
+      case LabelsRelation.name =>
+        new LabelsRelation(config)(sqlContext)
+      case SequenceRowsRelation.name =>
+        createSequenceRows(parameters, config, sqlContext)
+      case RelationshipsRelation.name =>
+        new RelationshipsRelation(config)(sqlContext)
+      case DataSetsRelation.name =>
+        new DataSetsRelation(config)(sqlContext)
+      case FlexibleDataModelBaseRelation.name =>
+        createFlexibleDataModelRelation(parameters, config, sqlContext)
+      case _ => sys.error(s"Resource type $resourceType does not support save()")
+    }
 
-        val operation: Seq[Row] => IO[Unit] = config.onConflict match {
-          case OnConflictOption.Abort =>
-            relation.insert
-          case OnConflictOption.Upsert =>
-            relation.upsert
-          case OnConflictOption.Update =>
-            relation.update
+    relation match {
+      case relation: DataPointsRelationV1[_] =>
```

Comment:
nit: while at prettifying maybe assets&datapoints could also be extracted into functions like createSequenceRows and createFlexibleDataModelRelation :thinking:

Reply:
There is definitely more to do here, but let's keep the scope limited